src/main/java/com/civclassic/oldenchanting/OldEnchanting.java

```diff
@@ -302,13 +306,41 @@ private void createXPBottles(Player player, int totalExp) {
 	
 	@EventHandler(priority = EventPriority.HIGHEST)
 	public void xpBottleEvent(ExpBottleEvent event) {
-		if(noExp) {
-			ProjectileSource source = event.getEntity().getShooter();
-			if(source instanceof Player) {
+		if (noExp) {
+			ThrownExpBottle bottle = event.getEntity();
+			ProjectileSource source = bottle.getShooter();
+			if (source instanceof Player) {
 				Player shooter = (Player) source;
 				shooter.giveExp(xpPerBottle);
+				bottle.teleport(shooter.getEyeLocation());
+			}
+			else if (source instanceof BlockProjectileSource) {
+				List<Player> nearby = bottle.getNearbyEntities(0.9, 0.9, 0.9)
+						.stream()
+						.filter((entity) -> entity instanceof Player)
+						.map((entity) -> (Player) entity)
+						.collect(Collectors.toList());
+				if (nearby.isEmpty()) {
+					event.setShowEffect(false);
+					event.setExperience(0);
+					return;
+				}
+				Player closest = null;
+				double distance = Double.MAX_VALUE;
+				for (Player player : nearby) {
```

Comment:
If you are already doing fancy stream stuff, you might have as well go all the way and do
```
.map((entity) -> (Player) entity)
.min(p ->  player.getLocation().distanceSquared(bottle.getLocation()))
```

which gives you an `Optional<Player>`, circumventing the need for both the collector and your for loop.

Reply:
In that case you could even do some obvious particle effect if the Optional is null to indicate that the XP is being wasted right now